### DISCUSSION TITLE
New option for migrations continue-on-failure

### DISCUSSION THREAD

**Discussion by pildit** [laravel/framework#40727] [Ideas]
Every time i got a migration error  I think about this : why there is not an option to continue-on-failure ?

**Comment by azankorejo** [comment]
@pildit I don't think that will make sense like what if the migration is has an dependency on the upcoming migration then it can be a disaster for developer to debug and plus database would be messed up.
And there are database errors too that should be addressed instead of brushing them off.

Keeping all other things in mind I believe that this in the end of the day brings complexity and more errors and depression :) into the developers life then making it easier.

But anyways thank you for putting your idea in here.

I hope you got your answer.

**Reply by pildit** [reply]
It's true what you said, but there are situations when I prefer to continue the migrations, like when the DB  already contains a big part of the migrations (but the migrations table is not synced with that). And there are other cases...that's why I proposed to be an option, not mandatory.

Other useful flag  would be something like --start-from=migrations_XXX so we could run all migrations from a point forward.

**Reply by oprypkhantc** [reply]
Why the hell would your DB already contain the migrations but not the actual records in `migrations`? It is your fault that you're doing DB modifications outside of Laravel migrations without rolling them back or creating a new DB.

There's something majorly wrong with your development process. Don't edit existing migrations. Don't do schema changes manually or roll them back - that's why rollbacks exist.

**Comment by ousid** [comment]
What type of errors do you get? Maybe [--force](https://laravel.com/docs/8.x/migrations#forcing-migrations-to-run-in-production) would help in your case?

**Reply by pildit** [reply]
Errors about columns already present in the table..I know i could modify the migrations  adding something like 
`if (! column_exists('new_column') ... ` 

but there are many migrations and i think there should be an easier way to just ignore this and continue.

**Reply by ousid** [reply]
I get this error many times, here is how to solve it.

First, there is an `up` method and `down` one in the `migration` class.

The `up` is for creating the `table`/`columns` and `down` for the opposite operation.

If you have duplicate column in your db, you need to do one of these options:

- Modify the existing column, see [this](https://dba.stackexchange.com/questions/169458/mysql-how-to-create-column-if-not-exists) (Not recommended)
- Rollback your database to previous migration to `undo` the previous migration operation or [use step flag](https://laravel.com/docs/8.x/migrations#rolling-back-migrations) to get the level you want (recommended)
- Create new migration to drop the column that is already exists on the DB (not sure why you do that, but still an option).

## In depth explanation for the 2nd option:

Let's say you have `articles` table migrated, and you need to add some column in it (e.g `excerpt`) You need to add a migration to add this one right? Something like: `XXX_add_excerpt_column_to_articles_table` and you will add the column to the migration file, BUT make sure to drop the column in the `down` method, because when you run `php artisan migrate:rollback` it will look at the `down` method first to drop the columns.

An example:
```php
    public function up()
    {
        Schema::table('articles', function (Blueprint $table) {
            $table->string('excerpt');
        });
    }

    /**
     * Reverse the migrations.
     *
     * @return void
     */
    public function down()
    {
        Schema::table('articles', function (Blueprint $table) {
          // when you rolling back the migration, laravel will look at this column/columns and drop.
            $table->dropColumn('excerpt');
        });
    }
```
Hope this answer helps you.

**Reply by pildit** [reply]
Thank you man, I already know this stuff...But i don't want to rollback, i have data in the DB which was not seeded so i don't want to lose it. WHY is it so hard to understand that SOMETIMES i just want that damn migration to continue ???  it's my choice if i do so!

**Reply by ousid** [reply]
Sorry to hear that @pildit, the only thing left then, is to extend the migration command logic and make it as you like.

Here:
https://github.com/laravel/framework/blob/8.x/src/Illuminate/Database/Console/Migrations/MigrateCommand.php

**Comment by TSVitor** [comment]
Same problem here, imported an old dump, wanna run the migration to get N new tables, dont care about all the "table already exist" errors just want laravel to add those that dont exist

**Comment by henzeb** [comment]
Had to deal with developers above me who would create a migration, and then did the modification manually on production, causing problems with CI. I always told them...check if the table already exists, check if it already has the columns, but they wouldn't or forgot. 

But I don't think forcing migrations to run anyway is the way to go. at least not in a way that it could be automated. you´d be carefully reviewing every exception. Did it fail because it already exists? or did it fail due some other error...

another aproach could be skipping specific migrations, but again, not automatable.

**Reply by VitorTS** [reply]
can be solved by adding specific flags like `--skip-if-table-already-exists`

**Comment by SilverPaladin** [comment]
the workaround would be to specify your migration classes with the ``` php artisan db:seed --class=MySeeder ``` option.

**Comment by shaedrich** [comment]
Usually, Laravel assumes, every migration can technically depend on any previously executed migration. So changing the order temporarily but not keeping a record of it might result in co-workers running into the same problem, something that Laravel strives to avoid.

However, before you commit a collection of multiple migrations, you sometimes need to test them and fiddle around with them to make them work. It might just as well be that they are not dependent on other migrations within said collection. To make your proposal work while keeping a record of skipping said migration, I suppose, it would be necessarily to add an option where the files are renamed to appear in a different order. So the failing migration then is moved behind the next migration and so  on or straight to the end. The reordered migrations would have to be run again to check if they work now before committing them.

**Comment by hassaans2008** [comment]
Hope the following blog will help you with the solution to go with: 
[Running Laravel Migrations While Skipping Runtime Errors: Your first Custom Laravel Command][1]


  [1]: https://medium.com/p/c6ab005fd928
